examples/persistent-counters/counter.go

```diff
@@ -0,0 +1,75 @@
+package main
+
+import (
+	"encoding/binary"
+	"fmt"
+	badger "github.com/dgraph-io/badger/v2"
+)
+
+type PersistentCounter struct {
+	db *badger.DB
+}
+
+func (pc *PersistentCounter) Get(key string) (int64, error) {
+	var res []byte
+	err := pc.db.View(func(txn *badger.Txn) error {
+		keyBytes := []byte(key)
+		item, err := txn.Get(keyBytes)
+		if err != nil {
+			return err
+		}
+		res, err = item.ValueCopy(nil)
+		return err
+	})
+	if err != nil {
+		return 0, err
+	}
+	val, _ := binary.Varint(res)
+	return val, nil
+}
+
+func (pc *PersistentCounter) Inc(key string, value int64) error {
+	presentVal, err := pc.Get(key)
+	if err != nil {
+		if err == badger.ErrKeyNotFound {
+			presentVal = 0
+		} else {
+			return err
+		}
+	}
+	err = pc.db.Update(func(txn *badger.Txn) error {
+		keyBytes := []byte(key)
+		valBytes := make([]byte, 64)
+		binary.PutVarint(valBytes, presentVal+value)
```

Comment:
do you need to do CAS manually. badger should support some form merge operator or maybe use transactions. Basically the reference code should be generic enough. CAS is mostly never possible.

Inc should add the key if it does not exist.